concepts-and-explanations/visitors.md

```diff
@@ -1,104 +1,114 @@
-* It implements Markable such that all LST elements can have [Markers](markers.md) applied to express additional meta-data about the element.
+### TreeVisitor
 
-### `TreeVisitor`
+All of OpenRewrite's visitors extend the abstract class `TreeVisitor<T extends Tree, P>`. It is this class that provides the generic, parameterized `visit(T, P)` method that drives a visitor's polymorphic navigation, cursoring, and life cycle.
 
-The framework provides the base class `TreeVisitor<T extends Tree, P>` from which all of OpenRewrite's visitors extend. It is this class that provides the generic, parameterized **`T visit(T, P)`** method that drives a visitor's polymorphic navigation, cursoring, and life-cycle. The parameterized type `T` represents the type of tree elements upon which the visitor will navigate and transform. The second parameterized type `P` is an additional, shared context that is passed to all visit methods as a visitor navigates a given LST.
+The parameterized type `T` represents the type of LSTs upon which the visitor will navigate and transform. The second paramter, `P`, is an additional, shared context that is passed to all visit methods as a visitor navigates a given LST (See [sharing data between visitors](#sharing-data-between-visitors) for more information).
 
 ### Cursoring
 
-All visitors have access to a `Cursor` which keeps track of a visitor's position within the LST while it is being traversed. Since LST elements are acyclic and therefore do not contain references to their parent element, the `Cursor` is the primary mechanism by which parent or sibling LST elements may be accessed during visiting. Logically a `Cursor` is a stack. Whenever an LST elment is visited a `Cursor` pointing to it is pushed on top of the stack. When visiting that element is over its `Cursor` is removed from the stack. In this way the `Cursor` keeps track of the visitor's current position within the LST.
+All visitors have access to a `Cursor` which keeps track of a visitor's position within the LST while it is being traversed. Since LSTs are acyclic and therefore do not contain references to their parent element, the `Cursor` is the primary mechanism by which parent or sibling LSTs may be accessed.
 
-As an example of how the `Cursor` can be helpful, imagine a visitor that is tasked with traversing a Java LST and marking only the top-level class as "final". The compilation unit may include a class that itself has several nested classes. Visiting such a tree would result in the `visitClassDeclaration()` method being called multiple times, once for each class declaration. The `Cursor` can be used to determine which class declaration represents the top-level class:
+Logically a `Cursor` is a stack. Whenever an LST is visited, a `Cursor` pointing to it is pushed on top of the stack. When the visit for the LST completes, its `Cursor` is removed from the stack. In this way, the `Cursor` keeps track of the visitor's current position within the LST.
```

Comment:
Would it maybe help to use _LST element_ to distinguish it from the LST as a whole? So something like:
```suggestion
Logically a `Cursor` is a stack. Whenever an LST element is visited, a `Cursor` pointing to it is pushed on top of the stack. When the visit for that LST element completes, its `Cursor` is removed from the stack. In this way, the `Cursor` keeps track of the visitor's current position within the LST.
```

Reply:
@sambsnyd and I have discussed this in the past. There was some confusion around the fact that `LST element` _is_ an `LST` vs. just a part of it. Because of this, it felt better to me to refer to everything as an `LST` and that `LSTs` are made up of other `LSTs`.

That being said, I appreciate the input and I'll keep that in mind going forward that the flip-side is confusing, too. Will think about if there's a better way.

Reply:
Makes sense. I checked https://docs.openrewrite.org/concepts-and-explanations/lossless-semantic-trees where LST element was being used.

Reply:
Good call out! Will update that to match as well. Thanks